ParabolInc/parabol
# Add string filtering to team / team member filter dropdowns

fixes https://github.com/ParabolInc/parabol/issues/5777.

Adds search functionality to the team / team member filter dropdowns.  This is done by creating a couple of new things:
* A `SearchMenuItem` component - a styled menu label that accepts text input
* A `useSearchFilter` hook - A hook that takes in an array of items and a function for getting a string value of those items, and returns:
  * An `onQueryChange` handler - Takes a query string.  Pass this as `SearchMenuItem`'s `onChange` handler, or any other text input.
  * `filteredItems` - An array of items that match the query string
  * `query` - The last query passed to `onQueryChange`.
  
These patterns are based on the [filtering logic](https://github.com/ParabolInc/parabol/blob/2d134896d2da154b34fc80a5025022a63cd35a4b/packages/client/components/GitHubScopingSearchFilterMenu.tsx#L145-L154)+[components](https://github.com/ParabolInc/parabol/blob/2d134896d2da154b34fc80a5025022a63cd35a4b/packages/client/components/GitHubScopingSearchFilterMenu.tsx#L172-L177) for integrations mentioned in the original ticket.

This also refactors several existing search filters to use this component+hook:
* TaskFooterIntegrateMenuList.tsx
* GitHubScopingSearchFilterMenu.tsx
* JiraScopingSearchFilterMenu.tsx
* NewGitHubIssueMenu.tsx
* NewJiraIssueMenu.tsx

(Each of these refactors have been tested with local GH + Atlassian integrations)

<img width="244" alt="Screen Shot 2022-01-26 at 2 40 51 PM" src="https://user-images.githubusercontent.com/9013217/151234898-edbe7e20-3355-4ba7-8bcc-83f64ae78d7a.png">

**Testing**
New search filtering:
* /me/tasks
  * Open the team dropdown (defaults to "All teams"), search for a team
  * Open the team member filter dropdown (defaults to current user), and search for a team member
  * Add a task to the board, click the task card's team, and search for a different team
* /team/:teamId
  * Open the team member filter dropdown ("all team members"), and search for a team member.  Confirm that team members are filtered and "All team members" is not shown when there's a search query.
  * Add a task card, click the task card's team member, and search for a different team member

Regression testing (for the refactors on existing search filtering):
* /me/tasks
  * Click the task card's "push to integration" button and search for GH and Atlassian integrations
* Sprint poker scoping
  * Filter by GH integration in scoping view
  * Filter by Jira integration in scoping view (search only shows up when more than 10 integrations are available)
  * For both Jira and GH, click new issue, and search for a different integration
